mlir/lib/Dialect/Rock/Transforms/GemmOutputSwizzle.cpp

```diff
@@ -0,0 +1,571 @@
+    int64_t mVectorLen = mVectorRes.max;
+    VectorizationResult nVectorRes =
+        getMaxVectorization(matC, dimensionN, /*inputDimLen=*/
+                            dataPerThread, matC.getDefiningOp());
+    int64_t nVectorLen = nVectorRes.max;
+    int64_t dim = (mVectorLen > nVectorLen) ? dimensionM : dimensionN;
+
+    // Get current workitem ID.
+    auto tid = b.create<WorkitemIdOp>(loc, b.getIndexType());
+
+    // Allocate LDS for output.
+    auto workgroupMemoryAddressSpace = b.getAttr<gpu::AddressSpaceAttr>(
+        gpu::GPUDialect::getWorkgroupAddressSpace());
+    auto ldsMemRefOutputType =
+        MemRefType::get({ldsRequiredBytes}, b.getI8Type(), AffineMap{},
+                        workgroupMemoryAddressSpace);
+    auto ldsBufferOutput = b.create<GpuAllocOp>(loc, ldsMemRefOutputType);
+    auto typedBuffer = viewBufferAs(b, ldsBufferOutput, destType);
+
+    // Convert from raw -> mPerBlock, nPerBlock
+    TopDownTMBuilder mnToRaw(b, {"gemmM", "gemmN"}, {mPerBlock, nPerBlock});
+    if (dim == dimensionN) {
+      mnToRaw.unmerge("flatten", 0, {"gemmM", "gemmN"}, {mPerBlock, nPerBlock});
+    } else {
+      mnToRaw.unmerge("flatten", 0, {"gemmN", "gemmM"}, {nPerBlock, mPerBlock});
+    }
+    auto mnToRawAttr = mnToRaw.get();
+
+    SmallVector<Attribute> transformMNToRawAttrs;
+    transformMNToRawAttrs.push_back(mnToRawAttr);
+    ArrayAttr transformMNToRaw = b.getArrayAttr(transformMNToRawAttrs);
+
+    auto ldsBufferMNToRaw = transform(b, typedBuffer, transformMNToRaw);
+
+    // LDS barrier.
+    // This barrier is needed because we reuse A and B buffers
+    b.create<LDSBarrierOp>(loc);
+
+    // Store C results to LDS.
+    b.create<ThreadwiseWriteAllOp>(loc, convertedC, ldsBufferMNToRaw,
+                                   /*extraViews=*/idToLDS,
+                                   /*extraIndices=*/ValueRange{tid},
+                                   op.getFeatures(), StoreMethod::Set,
+                                   /*forceUnroll=*/forceUnroll,
+                                   /*useIndexDiffs=*/useIndexDiffs);
+
+    // Load from LDS to registers.
+    int64_t elementsWrittenPerThread = 128 / destType.getIntOrFloatBitWidth();
+    elementsWrittenPerThread =
+        math_util::gcd(dataPerThread, elementsWrittenPerThread);
+    int64_t iter = dataPerThread / elementsWrittenPerThread;
+    LLVM_DEBUG(llvm::dbgs()
+               << "blockSize: " << blockSize
+               << " dataPerThread: " << dataPerThread
+               << " elementsWrittenPerThread: " << elementsWrittenPerThread
+               << " iter: " << iter << "\n");
+    if (dim == dimensionM) {
+      LLVM_DEBUG(llvm::dbgs() << "dim = M\n");
+    } else {
+      LLVM_DEBUG(llvm::dbgs() << "dim = N\n");
+    }
+
+    Value finalC =
+        gpuAlloc(b, loc, dataPerThread, destType, AddressSpace::Private);
+
+    TopDownTMBuilder tidIterMerge(b, {"tid", "iter"},
+                                  {blockSize, dataPerThread});
```

Comment:
So yea -- `dataPerThread` should be extracted from original `threadwise_write_all` for stability of this